client/signup/main.jsx

```diff
@@ -314,29 +326,53 @@ class Signup extends React.Component {
 	handleFlowComplete = ( dependencies, destination ) => {
 		debug( 'The flow is completed. Destination: %s', destination );
 
-		const isNewUser = !! ( dependencies && dependencies.username );
+		const { isNewishUser, existingSiteCount } = this.props;
+
+		const isNewUser = !! (
+			( dependencies && dependencies.username ) ||
+			( isNewishUser && dependencies && dependencies.siteSlug && 0 === existingSiteCount - 1 )
+		);
 		const isNewSite = !! ( dependencies && dependencies.siteSlug );
 		const hasCartItems = !! (
 			dependencies &&
 			( dependencies.cartItem || dependencies.domainItem || dependencies.themeItem )
 		);
-		const isNewUserOnFreePlan = isNewUser && isNewSite && ! hasCartItems;
 
-		analytics.recordSignupComplete( {
+		const debugProps = {
+			isNewishUser,
+			existingSiteCount,
 			isNewUser,
 			isNewSite,
 			hasCartItems,
-			isNewUserOnFreePlan,
 			flow: this.props.flowName,
-		} );
+		};
+		debug( 'Tracking signup completion.', debugProps );
 
-		if ( dependencies.cartItem || dependencies.domainItem ) {
-			this.handleLogin( dependencies, destination );
-		} else {
-			this.setState( {
-				loginHandler: this.handleLogin.bind( this, dependencies, destination ),
-			} );
-		}
+		this.setState( { isTrackingSignup: true } );
+		this.updateShouldShowLoadingScreen();
+
+		analytics.recordSignupComplete(
+			{
+				isNewUser,
+				isNewSite,
+				hasCartItems,
+				flow: this.props.flowName,
+			},
+			() => {
+				debug( 'Done tracking signup completion.' );
+
+				this.setState( { isTrackingSignup: false } );
+				this.updateShouldShowLoadingScreen();
+
+				if ( dependencies.cartItem || dependencies.domainItem ) {
```

Comment:
@scruffian @jsnmoon On this change in particular, I would really appreciate a review from someone who is more familiar with the signup lib in Calypso. On signup complete, we're using a delay of 1000-3000 ms when tracking signups. There are several ad trackers and we need to give them time to load before redirecting.

Can you take a look at the changes in this file and let us know if you have recommendations?

Reply:
@adamziel Is this something you could help with?
I'm looking for another pair of eyes on the proposed changes in this file.

Reply:
I'm not a big fan of this approach:

Wrapping an action to happen inside analytics is a bit bad because if someone has some kind of an adblock/tracking block it might break the flow. 

Is there a requirement to continue only after tracking the Signup completion?

If not, could this be rewritten in a way to spawn off the action in parallel while the flow continues forward? 

Is moving the login code as a callback a result of the above question about loading external tracking libraries? Because the current way it works is it fires of the events asynchronously and continues the flow.

As for the artificial delay of 1 to 3 seconds to load the libraries - this is very error-prone approach. On slower connections this period might be far from enough to make external requests and load everything.

A few ways to make this work better would be to:

* Queue loading the ad tracking libraries as early as possible in the signup flow. This way when the customer reaches the end it will have had time to load the libraries even on slower connections. 
* Store the trackings we want to do and do them after the user has logged in Calypso. For example in `localStorage`. This would work well if we already load those tracking libraries in Calypso.
* Try best effort tracking - this will surely lose some date - fire off the action and hope it completes before the user navigates away from the page.

Reply:
@bisko  Thanks for digging into this and taking a look. 🙏

> Is there a requirement to continue only after tracking the Signup completion?

Not to wait on the trackers to load or finish, only that we wait to continue until signup tracking has been _attempted_. In a case when tracking is being attempted, that we wait up to 3 seconds for the trackers to record what they need to.

> If not, could this be rewritten in a way to spawn off the action in parallel while the flow continues forward?

This is the approach we took previously and there has been some doubt about whether it works effectively or not, because this results in a race condition. If signup marches on and does a full page change (e.g., moving the user to checkout, or to their site) then it could occur before the trackers have done their thing.

> [...] artificial delay of 1 to 3 seconds to load the libraries [...] 
> this is very error-prone approach

Agree. Thanks for calling that out.

> Store the trackings we want to do and do them after the user has logged
> in Calypso. For example in localStorage.

That idea could work. Thanks! Let me take a stab at that and see what I can do.

Reply:
@bisko wrote...

> Store the trackings we want to do and do them after the user has logged in Calypso. For example in localStorage. This would work well if we already load those tracking libraries in Calypso.

This works quite well. Thanks again! A new `analytics.queue.add()` and a family of functions have been added to this PR. So now instead of `setTimeout` I'm using `localStorage` to queue `calypso_signup_complete` and the other ad-tracking events that runs along with it.

cc @lsinger @micbosi 

Here's what to expect when testing this portion again (screenshot below). Please be sure to set `flags` as a cookie so they are persisted across pageview events.

```
document.cookie = 'flags=gdpr-banner,google-analytics,ad-tracking';
```

![Screen Shot 2019-04-21 at 21 55 23](https://user-images.githubusercontent.com/1563559/56485783-dd6d5700-6481-11e9-8bb6-b63b8aa651da.png)

Reply:
In my testing this worked as intended -- I think we should merge this today. :shipit: